### PR TITLE
Timeline: Adds opacity & line width option

### PR DESCRIPTION
**Pull request by torkelo**
* [x] Minor refactorings to improve code readability, but struggling a lot with how to write tests for the timeline code and how to reduce nesting 
* [x] Adds fillOpacity option and lineWidth 
* [x] Tested a lot with a gradient modes but none impressed so skipping that option for now 
* [x] Updates the test dashboard 

 There is an issue with getContastText where it does not take into account opacity colors, created issue for that: https://github.com/grafana/grafana/issues/34117 

There is also an issue with line width does not actually work for "state changes" mode,  strokePaths map only contains style (color) and path not lineWidth, so in
https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/timeline/timeline.ts#L114 there is no access to stokeWidth 

Screenshot from 2021-05-14 12-58-23

### CODE REVIEW

File: packages/grafana-data/src/utils/OptionsUIBuilders.ts [hunk 1]
```diff
@@ -33,7 +32,7 @@ export class FieldConfigEditorBuilder<TOptions> extends OptionsUIRegistryBuilder
       override: standardEditorsRegistry.get('number').editor as any,
       editor: standardEditorsRegistry.get('number').editor as any,
       process: numberOverrideProcessor,
-      shouldApply: config.shouldApply ? config.shouldApply : (field) => field.type === FieldType.number,
+      shouldApply: config.shouldApply ?? (() => true),
```
Comment (torkelo) [inline packages/grafana-data/src/utils/OptionsUIBuilders.ts:35]: we had this strange default, where field options by default only applied to number fields. Removing this (potentially breaking change)

**Comment by leeoniya** [comment]
@torkelo please revert the "State changes" panels to their original, manually-entered timestamps and values, and the dashboard to a pre-defined time range. these settings were intentionally chosen to show what real data will look like for this mode (it has to be misaligned, and shows precisely how it is different from the "Periodic samples" mode. the modifications destroy this and make both look the same.

**Comment by torkelo** [comment]
ok, will create a new test dashboard then

**Comment by torkelo** [comment]
@leeoniya pushed an update that adds back the original test dashboard but with the proper dashboard tags so it shows up in the Panel tests lists. 

think I spot a bug in the first panel on that dashboard, the first 0 value is not being rendered (same on master so not due to this PR) 
Screenshot from 2021-05-14 15-06-42

**Comment by leeoniya** [comment]
> think I spot a bug in the first panel on that dashboard, the first 0 value is not being rendered (same on master so not due to this PR)

yes we still need to figure out what to do here. the data that will come back will always be wider than the dash time range so that we can render the tail ends of previous states. how this time expansion happens on the datasource side is not yet clear, but those state labels should probably be rendered right-aligned, or something.

**Comment by torkelo** [comment]
@leeoniya ah, now I understand the data point is before the time range. And yea we have the same problem with the last data point is just before time range end, we should expand the time range so we can show that data point & value, but tricky

**Comment by leeoniya** [comment]
this is another reason why i think in "State changes" mode, center-aligned text makes no sense. it's essentially a trace view:

image

**Comment by torkelo** [comment]
@leeoniya in many many cases the data will be periodic and aligned, and in those cases it makes sense

**Comment by leeoniya** [comment]
in periodic/aligned mode, the text will always be centered by default anyways and in state changes (traces) mode i dont think it makes sense to have different alignment except for state start.

**Comment by torkelo** [comment]
but I agree in many cases left-aligned will be the right choice 

> in periodic/aligned mode, the text will always be centered 

Yes, but when the data is aligned, why do I have to set an option to also display it as aligned? There might be cases where you have 2-4 series that have periodic data but not aligned, but you do not want the gaps that periodic mode creates, then you are forced to live with left-aligned text if you want no gaps. (if there where not alignment option) 

for example this would not be possible
Screenshot from 2021-05-14 16-32-32
you would have to live with this if you wanted centered text 
Screenshot from 2021-05-14 16-32-54

**Comment by leeoniya** [comment]
"living with it" is an odd way to view this. in periodic mode the width of the boxes is meaningless, it does not represent the duration of that sample. this is a key distinction. periodic mode should not fill the space just because it looks nice. these are instantaneous events. showing periodic sample data with different width boxes is misleading for the sake of it looking nice, imo.

**Comment by torkelo** [comment]
> showing periodic sample data with different width boxes is misleading for the sake of it looking nice, imo.

But I am not arguing for that. In the screenshot below you have 4 series, two with different periods. I want the boxes to go up to the next periodic data point for that series, and "state changes" is the only mode that does that right now. Maybe this is a misuse of that mode and we should have a periodic mode that supports different periods for each series?  

Screenshot from 2021-05-14 16-32-54

**Comment by leeoniya** [comment]
> Maybe this is a misuse of that mode and we should have a periodic mode that supports different periods for each series?

the above sample looks okay-ish when you have 2 sets of aligned series adjacently, but if you just take B and C, now it will look like the box widths represents duration. imo, the boxes need to be as narrow as possible to contain the text with padding. expanding them along x implies temporal semantics that do not exist in "samples" mode.

i do think that the axis colliding ticks/labels issue needs to be addressed for periodic mode where you can have differently-periodic/mis-aligned series, but that doesnt touch the width consistency.

FWIW, i do think yours looks nicer, but what it's showing is not clear in all cases.

**Pull request merged**